compositor_pipeline/src/queue.rs

```diff
@@ -0,0 +1,97 @@
+mod internal_queue;
+
+use std::{
+    sync::{Arc, Mutex},
+    thread::{self, sleep},
+    time::{Duration, Instant},
+};
+
+use compositor_common::{Frame, Framerate, InputId};
+use compositor_render::input_frames::InputFrames;
+use crossbeam_channel::{tick, unbounded, Receiver, Sender};
+
+use self::internal_queue::{InternalQueue, QueueError};
+
+const DEFAULT_BUFFER_DURATION: Duration = Duration::from_millis(10);
+
+pub struct Queue {
+    internal_queue: Arc<Mutex<InternalQueue>>,
+    check_queue_channel: (Sender<()>, Receiver<()>),
+    output_framerate: Framerate,
+    buffer_duration: Duration,
+}
+
+impl Queue {
+    pub fn new(output_framerate: Framerate) -> Self {
+        Queue {
+            internal_queue: Arc::new(Mutex::new(InternalQueue::new(output_framerate))),
+            check_queue_channel: unbounded(),
+            output_framerate,
+            buffer_duration: DEFAULT_BUFFER_DURATION,
+        }
+    }
+
+    pub fn add_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        internal_queue.add_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
+        internal_queue.remove_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn start(&self, sender: Sender<InputFrames>) {
+        let (check_queue_sender, check_queue_receiver) = self.check_queue_channel.clone();
```

Comment:
you only use that channel for ticker, but it should also send event after enqueue

Reply:
Are you sure? I think that cloned `crossbeam channel` will work as indented here, based on [this example](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Sender.html#examples). 

I modified the code a little bit to check that and according to logs it should work.
```rust
#[derive(Debug)]
enum Event {
    Tick,
    Enqueue
}

pub struct Queue {
    internal_queue: Arc<Mutex<InternalQueue>>,
    check_queue_channel: (Sender<Event>, Receiver<Event>),
    output_framerate: Framerate,
    buffer_duration: Duration,
}

impl Queue {
    ...

    pub fn start(&self, sender: Sender<InputFrames>) {
        let (check_queue_sender, check_queue_receiver) = self.check_queue_channel.clone();
        let internal_queue = self.internal_queue.clone();
        let tick_duration = self.output_framerate.get_interval_duration();
        let buffer_duration = self.buffer_duration;

        thread::spawn(move || {
            // Wait for first frame
            check_queue_receiver.recv().unwrap();
            Self::start_ticker(tick_duration, check_queue_sender);

            let start = Instant::now();
            loop {
                let event = check_queue_receiver.recv().unwrap();
                info!("Render loop {:#?}", event);

                let mut internal_queue = internal_queue.lock().unwrap();
                let buffer_pts = internal_queue.get_next_output_buffer_pts();

                if start.elapsed() + buffer_duration > buffer_pts
                    || internal_queue.check_all_inputs_ready(buffer_pts)
                {
                    let frames_batch = internal_queue.get_frames_batch(buffer_pts);

                    sender.send(frames_batch).unwrap();

                    internal_queue.drop_useless_frames();
                }
            }
        });
    }

    pub fn enqueue_frame(&self, input_id: InputId, frame: Frame) -> Result<(), QueueError> {
        let mut internal_queue = self.internal_queue.lock().unwrap();

        internal_queue.enqueue_frame(input_id, frame)?;
        internal_queue.drop_pad_useless_frames(input_id)?;

        self.check_queue_channel.0.send(Event::Enqueue).unwrap();

        Ok(())
    }

    fn start_ticker(tick_duration: Duration, check_queue_sender: Sender<Event>) {
        thread::spawn(move || {
            let ticker = tick(tick_duration);
            loop {
                ticker.recv().unwrap();
                check_queue_sender.send(Event::Tick).unwrap();
            }
        });
    }
}
```

logs:
```rust
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Enqueue
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Enqueue
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Enqueue
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:13:59Z INFO  compositor_pipeline::queue] Render loop Enqueue
[2023-07-24T19:14:00Z INFO  compositor_pipeline::queue] Render loop Tick
[2023-07-24T19:14:00Z INFO  compositor_pipeline::queue] Render loop Enqueue
[2023-07-24T19:14:00Z INFO  compositor_pipeline::queue] Render loop Tick

```

Reply:
ahh sorry, I was checking where check_queue_sender is passed and did not realize that you are cloning it there not creating

Reply:
I don't really like the way the queue is stored and copied, but I don't have any better idea on how to handle it.